src/hotspot/share/gc/shared/slidingForwarding.inline.hpp

```diff
@@ -0,0 +1,137 @@
+  size_t from_reg_idx = region_index_containing(from);
+  size_t to_reg_idx = region_index_containing(to);
+
+  HeapWord* to_region_base = _heap_start + to_reg_idx * _region_size_words;
+  size_t base_idx = from_reg_idx * NUM_TARGET_REGIONS;
+
+  bool alt_region = false;
+  if (_bases_table[base_idx] == UNUSED_BASE) {
+    // Primary is free
+    _bases_table[base_idx] = to_region_base;
+  } else if (region_contains(_bases_table[base_idx], to)) {
+    // Primary is good
+  } else {
+    size_t base_idx_alt = base_idx + 1;
+    if (_bases_table[base_idx_alt] == UNUSED_BASE) {
+      // Alternate is free
+      _bases_table[base_idx_alt] = to_region_base;
+    } else if (region_contains(_bases_table[base_idx_alt], to)) {
+      // Alternate is good
+    } else {
+      // Both primary and alternate are not fitting
+      assert(UseG1GC, "Only happens with G1 serial compaction");
+      return (1 << FALLBACK_SHIFT) | markWord::marked_value;
+    }
+    alt_region = true;
+  }
+
+  size_t offset = pointer_delta(to, to_region_base);
+  assert(offset < _region_size_words, "Offset should be within the region. from: " PTR_FORMAT
+         ", to: " PTR_FORMAT ", to_region_base: " PTR_FORMAT ", offset: " SIZE_FORMAT,
+         p2i(from), p2i(to), p2i(to_region_base), offset);
+
+  uintptr_t encoded = (offset << OFFSET_BITS_SHIFT) |
+                      (alt_region << ALT_REGION_SHIFT) |
+                      markWord::marked_value;
+
+  assert(to == decode_forwarding(from, encoded), "must be reversible");
+  return encoded;
+}
+
+HeapWord* SlidingForwarding::decode_forwarding(HeapWord* from, uintptr_t encoded) const {
+  assert((encoded & markWord::marked_value) == markWord::marked_value, "must be marked as forwarded");
+  size_t alt_region = (encoded >> ALT_REGION_SHIFT) & right_n_bits(ALT_REGION_BITS);
+  assert(alt_region < NUM_TARGET_REGIONS, "Sanity");
+  uintptr_t offset = (encoded >> OFFSET_BITS_SHIFT);
+
+  size_t from_idx = region_index_containing(from) * NUM_TARGET_REGIONS;
+  size_t base_idx = from_idx + alt_region;
+
+  HeapWord* decoded = _bases_table[base_idx] + offset;
+  assert(decoded >= _heap_start,
+         "Address must be above heap start. encoded: " INTPTR_FORMAT ", alt_region: " SIZE_FORMAT ", base: " PTR_FORMAT,
+         encoded, alt_region, p2i(_bases_table[base_idx]));
+
+  return decoded;
+}
+
+void SlidingForwarding::forward_to(oop from, oop to) {
+  assert(_bases_table != nullptr, "call begin() before forwarding");
+
+  markWord from_header = from->mark();
+  if (from_header.has_displaced_mark_helper()) {
+    from_header = from_header.displaced_mark_helper();
+  }
+
+  HeapWord* from_hw = cast_from_oop<HeapWord*>(from);
+  HeapWord* to_hw   = cast_from_oop<HeapWord*>(to);
+  uintptr_t encoded = encode_forwarding(from_hw, to_hw);
+  markWord new_header = markWord((from_header.value() & ~MARK_LOWER_HALF_MASK) | encoded);
+  from->set_mark(new_header);
```

Comment:
When the header is displaced, it will be recorded in the preserved-marks table. Then we over-write the mark-word with the forwarding. At the end of the GC, we will restore the original mark from the preserved-marks table. This is the same mechanism that is already used in normal uncompressed forwarding.